sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuMapUtils.scala

```diff
@@ -0,0 +1,161 @@
+          if (idx != index) {
+            childViews(idx).safeClose()
+          }
+        }
+      }
+    }
+  }
+
+  private[this] def replaceExplodedKeyOrValueAsView(
+      mapCol: ColumnView,
+      toReplace: ColumnView,
+      index: Int): ColumnView = {
+    withResource(mapCol.getChildColumnView(0)) { keyValueView =>
+      withResource(replaceStructChild(keyValueView, toReplace, index)) { newKeyValueView =>
+        GpuListUtils.replaceListDataColumnAsView(mapCol, newKeyValueView)
+      }
+    }
+  }
+
+  /**
+   * Replace the values in a map. The values are the underlying exploded values (not in a list)
+   * @param mapCol the original map column
+   * @param newValue the new values column
+   * @return and updated map column view
+   */
+  def replaceExplodedValueAsView(
+      mapCol: ColumnView,
+      newValue: ColumnView): ColumnView =
+    replaceExplodedKeyOrValueAsView(mapCol, newValue, VALUE_INDEX)
+
+  /**
+   * Replace the keys in a map. The values are the underlying exploded keys (not in a list)
+   * @param mapCol the original map column
+   * @param newKey the new keys column
+   * @return and updated map column view
+   */
+  def replaceExplodedKeyAsView(
+      mapCol: ColumnView,
+      newKey: ColumnView): ColumnView =
+    replaceExplodedKeyOrValueAsView(mapCol, newKey, KEY_INDEX)
+
+  def assertNoNullKeys(mapView: ColumnView): Unit = {
+    withResource(mapView.getChildColumnView(0)) { keyValueList =>
+      withResource(keyValueList.getChildColumnView(KEY_INDEX)) { keyView =>
+        if (keyView.getNullCount > 0) {
+          throw new RuntimeException("Cannot use null as map key.")
+        }
+      }
+    }
+  }
+
+  // Copied from Spark org.apache.spark.sql.errors.QueryExecutionErrors
+  private[this] def duplicateMapKeyFoundError: Throwable = {
+    new RuntimeException(s"Duplicate map key was found, please check the input " +
+        "data. If you want to remove the duplicated keys, you can set " +
+        s"${SQLConf.MAP_KEY_DEDUP_POLICY.key} to ${SQLConf.MapKeyDedupPolicy.LAST_WIN} so that " +
+        "the key inserted at last takes precedence.")
+  }
+
+  /**
+   * In Spark there are two options to deal with duplicate keys in a map. The default is to throw
+   * an exception. This can handle that use case.
+   * @param mapView the view to check to see if it has duplicate keys or not.
+   */
+  def assertNoDuplicateKeys(mapView: ColumnView): Unit = {
+    val dedupeCountCol = withResource(getKeysAsListView(mapView)) { keyListView =>
+      withResource(keyListView.dropListDuplicates) { dupsDropped =>
+        dupsDropped.countElements()
+      }
+    }
+    val sameLengthCol = withResource(dedupeCountCol) { dedupeCountCol =>
+      withResource(mapView.countElements()) { origCount =>
+        origCount.equalToNullAware(dedupeCountCol)
+      }
+    }
+    withResource(sameLengthCol) { sameLengthCol =>
+      withResource(sameLengthCol.all()) { allSame =>
+        if (!allSame.getBoolean) {
```

Comment:
Annoyingly, `all()` can produce an invalid scalar in some cases.
```suggestion
        if (allSame.isValid && !allSame.getBoolean) {
```

Reply:
That is why I did `equalToNullAware` instead of `equalTo`. It cannot produce a null output, but I think you are right. If there are no rows it still produces null.